### ISSUE TITLE
Triton + Tensorflow: Batching issues - Latency per input constant regardless of batch size

### ISSUE THREAD

**Issue by NiklasA11** [triton-inference-server/server#6894]
I am new to Triton, and am trying out some Tensorflow models I have laying around to see if it is worth porting to Triton from Tensorflow Serving to allow us to support more than just Tensorflow models.

So, I have a couple of Tensorflow models that all have inputs that have the first axis as the batch axis. The models thus accept a variable number of inputs and can perform operations on the full batch in one go.

When running the model in Tensorflow Serving or just in a python script, if I pass an input tensor with e.g. 32 elements, the model will process that faster compared to sending 32 individual requests of size 1.  The latency "per request" should go down when increasing the batch size, since the model can perform calculations in parallell when it works on batches of requests.
 
**The problem** however is that I get absolutely no speedup when sending batched requests to the Triton Server.

Note: I don' think this is related to the dynamic batching settings, since as I understood it, dynamic batching is about combining data from multiple requests that come to Triton, and then send all of them in one request to the underlying model. 

This is about that when I send sequential requests, with the batch axis>1, for Triton, I get a more or less constant increase. So, 1 request with 16 inputs takes roughly 16x the time for a request with 1 input when sending requests to Triton. This is not the case when querying the model directly from Python, nor when serving the model using Tensorflow Serving.
 
I have benchmarked Triton against Tensorflow Serving. For each batch size,  I am making 5 sequential request with different and print the total number taken for the request, as well as the average time per input text across all the texts/requests. The pattern below is the same for all models, so I will just print one example here. Since it's the same pattern for all models, I also don't think the issue is with my actual model, but rather something with the way I am using Triton or setting up the config.

This particular model is a CNN Text classifier with a single text string as input and a list of tensors (label, label score, label certainty, etc) as output.

**Tensorflow Serving**
When calling the Tensorflow Serving REST API that is serving the model , I get the following response times:
```
niklas.andersson@C02GC15RMD6R serving % python3 test_tf_model.py
Total time for 5 requests with  batch size 1:            0.17203783988952637 seconds    Total per text: 0.03440756797790527 seconds
Total time for 5 requests with  batch size 2:            0.17231225967407227 seconds    Total per text: 0.017231225967407227 seconds
Total time for 5 requests with  batch size 4:            0.1922588348388672 seconds     Total per text: 0.009612941741943359 seconds
Total time for 5 requests with  batch size 8:            0.19364094734191895 seconds    Total per text: 0.004841023683547973 seconds
Total time for 5 requests with  batch size 16:           0.20978665351867676 seconds    Total per text: 0.0026223331689834593 seconds
Total time for 5 requests with  batch size 32:           0.3673431873321533 seconds     Total per text: 0.002295894920825958 seconds


```

We see that the latency per text is going by a factor 15 when increasing batch size from 1 to 32, because Tensorflow is more efficient at processing a few larger batches compared to processing many small batches. This is what you would expect since there are operations in the network that benefit from working on batches instead of one item at a time.

So far so good.

**Triton**
When running the same test for Triton, (the same model),  I get the following stats:

```
Time for 5 requests or batch size 1:           1.131458044052124 seconds  total per text: 0.22629160881042482 seconds
Time for 5 requests or batch size 2:           1.7335660457611084 seconds, total per text: 0.17335660457611085 seconds
Time for 5 requests or batch size 4:           3.85111927986145 seconds, total per text: 0.1925559639930725 seconds
Time for 5 requests or batch size 8:           6.531412839889526 seconds, total per text: 0.16328532099723816 seconds
Time for 5 requests or batch size 16:         12.642588138580322 seconds, total per text: 0.15803235173225402 seconds
Time for 5 requests or batch size 32:         26.28559708595276 seconds, total per text: 0.16428498178720474 seconds
```

We can see something is very off. Triton, for the same model,  needs 26 (!!!) seconds for 5 batches with 32 texts, while the same model on Tensorflow Serving serves 5 batches of size 32 in 0.367 seconds. Or, per text, Triton takes on average 0.16428498178720474 seconds per text when using batch size 32, while Tensorflow Serving takes 0.002295894920825958 seconds. That's a factor 71 difference!!!

But even more concerning, the _**batch size seems to have zero effect**_ when using Triton. The average time per text for Triton across is constant regardless of the batch size I send in the request. This makes me think that Triton is not batching the requests in any way at all. On the contrary, it seems that Triton "un-batches" them in the sense that even if I send one request to the Triton API with 32 entries, **_Triton still appears to be sending them one at a time to the underlying Tensorflow model_**. This would explain the constant time per text and the linear increase in time.

Also, it is a bit worrying that even for batch size 1, Tensorflow is averaging 0.03440756797790527 per text, while Triton is averaging 0.22629160881042482 per text, so Triton is a factor 6 slower than Tensorflow Serving for batch size 1, in addition to not scaling well at all for batched requests. 

I really hope I am doing something wrong in my model config, because this makes absolutely zero sense to me :-) 

I do not believe that the issue is related to the model, since the behavior is the same for all of my own models I have tried as well as for open source models I fetched from the internet. Also, calling the model directly from Python and through Tensorflow Serving shows the expected behavior, so I really believe something is off in the way I have configured Triton, or there is some problem with Triton itself.

My model config in Triton looks like follows: 

```
name: "model"
platform: "tensorflow_savedmodel"


dynamic_batching {
    max_queue_delay_microseconds: 100
}

max_batch_size: 512

input: [
    {
      name: "text_input"
      data_type: TYPE_STRING
      dims: [1]
    }
]

output: [ ... all my outputs ...]
instance_group: [
    {
      count: 4
      kind: KIND_CPU
    }
]
```

I have also tried to not specifying the inputs and outputs, because Triton can derive those fields from the Tensorflow model. I have also tried explicitly adding the -1-axis to the inputs/outputs in the config. It all gives the same result.

One request I sent to the Triton http server looks like this (batch size 4):
`{"inputs":[{"name":"text_input","shape":[4,1],"datatype":"BYTES","data":[["text 1","text 2","text 3","text 4"]]}]}
`
Please advise, because this makes zero sense to me.

On the positive note, it looks like for batch size 1, the requests take roughly the same amount of time, which means that no "silly" things are done with the actual Tensorflow runtime, i.e. running one input through the model takes roughly the same time regardless of whether I am using Triton or Tensorflow Serving. The difference is that Triton seems to send all the inputs one at the time to the model, so we never get any speedup from batch processing.

Many thanks in advance for any pointers what so ever.

Best regards,
Niklas

P.S. For reference, I am creating my Dockerfile in this way because I need the Tensorflow Sentencepiece module in this particular model. Again, the behavior above is the same even for models that do not use the Sentencepiece module.

```
FROM nvcr.io/nvidia/tritonserver:24.01-py3 

RUN python3 -m pip install tensorflow-text==2.14.0
ENV LD_LIBRARY_PATH=/opt/tritonserver/backends/tensorflow:$LD_LIBRARY_PATH
ENV LD_PRELOAD=/usr/local/lib/python3.10/dist-packages/tensorflow_text/python/ops/_sentencepiece_tokenizer.so

ENTRYPOINT ["tritonserver"]

```
P.S. 2: I believe the models are set up correct to handle batches. Looking at the numbers from Tensorflow Serving, that appears to be the case, but just for clarity, this is the way the Tensorflow model was exported. Note the [None, 1]-shape, which indicates that the first axis is the batch axis of unknown shape, and the second dimension has shape 1.

```python
export_signature = model.call.get_concrete_function(
        text_input=tf.TensorSpec(shape=[None, 1], dtype=tf.dtypes.string, name="text_input")
    )

    tf.keras.models.save_model(
        serving_model,
        output_model_path,
        signatures={
            "predict_signature": export_signature
        }
    )

```

**Comment by NiklasA11** [comment]
I have now tried the same on a EC2 setup, thinking maybe something was off on my local machine.

Although the exact numbers differ, the trend is the same. Triton has a constant "time per text" regardless of the batch size, while Tensorflow Serving benefits from receiving batches (i.e. the "time per text" goes down when sending larger batches).

For Triton, the pattern is also the same regardless of whether I run on a CPU or a GPU.

Does anyone have any advise whatsoever on what to look for?

**Comment by jalola** [comment]
It is strange to me that Triton taking longer time to run inference.
Under the hood, I think it uses Tensorflow to run inference then the performance should be similar to Tensorflow serving.

Can you use performance analyzer to check the throughput and latency instead of your script?
https://github.com/triton-inference-server/client/blob/main/src/c++/perf_analyzer/README.md

**Comment by NiklasA11** [comment]
Thank you very much for your response!

perf tool output:
`
root@docker-desktop:/workspace# perf_analyzer -m test-model-1
*** Measurement Settings ***
  Batch size: 1
  Service Kind: Triton
  Using "time_windows" mode for stabilization
  Measurement window: 5000 msec
  Using synchronous calls for inference
  Stabilizing using average latency

Request concurrency: 1
  Client: 
    Request count: 93
    Throughput: 5.1611 infer/sec
    Avg latency: 193902 usec (standard deviation 17085 usec)
    p50 latency: 187210 usec
    p90 latency: 218210 usec
    p95 latency: 227079 usec
    p99 latency: 250003 usec
    Avg HTTP time: 193890 usec (send/recv 89 usec + response wait 193801 usec)
  Server: 
    Inference count: 93
    Execution count: 93
    Successful request count: 93
    Avg request latency: 193398 usec (overhead 66 usec + queue 11067 usec + compute input 40 usec + compute infer 182183 usec + compute output 42 usec)

Inferences/Second vs. Client Average Batch Latency
Concurrency: 1, throughput: 5.1611 infer/sec, latency 193902 usec
`
5.1611 request per second = 193ms per request, so in the same ballpark figure as I got from my script.

In addition to my own script, I have also benchmarked the service with Locust and the linux ab command and I see numbers that are really close to this (when running concurrency 1 in locust and "ab", obviously).

**Comment by NiklasA11** [comment]
but as I see it, there are two problems. Maybe they are related,  maybe not.

- For one single request with one single input, I don't expect Triton to be around 6 times slower than Tensorflow Serving. In this case, there's nothing in the dynamic batching settings really that I believe could explain this, so the issue must be something else?

- For requests with multiple inputs, it appears from the timings in my experiment that Triton is looping over the inputs and sending them to the model one at a time. As I understood, the dynamic baching is about grouping together multiple different requests, so this is maybe also not related to the dynamic batching, but something else?

**Comment by jalola** [comment]
I think you should focus on the first point to check why Triton is slower than TF Serving 6 times on batch 1 and much more on bigger batch
I know Triton just uses Tensorflow backend, same as TF Serving
Can you load the TF model and run model.predict in Python code without Triton or TF Serving to see the latency of different batches?
Another thing you can check whether they produce the same output from the same input. Just to have a sanity check that you are benchmarking on the same model for both Triton and TF Serving

I just ran similar comparison between Triton and TF Serving, they shows similar performance.
In my case, it is another way around, Triton and TF Serving are fast but the without serving (loading model by python --> run inference) is very slow. I still didn't find the reason for that yet

Another idea can be sentencepiece_tokenizer is super slow on Triton. What if you cut this layer out of the model inference, will the Triton and TF Serving have the same performance? I don't know :)

**Comment by NiklasA11** [comment]
I created a dummy model with just a couple of convolutional layers and a couple of tense layers, and indeed, the performance is much more in line with what you would expect.

Maybe the senteice_piece-tokeniser is the "bad guy" here. It's kind of hard to just cut it out because then I need to change my inputs to be something else than strings, etc, so it's nothing I can do in 5 minutes, but I will for sure give it a go.  I see there are multiple different sentence-piece tokenisers available in TF/Keras. I will try to use on of the other ones instead and see if that fixes it.

I have a couple of lookup-layers in the model as well, I will check if maybe those are problematic for Triton to handle.

But it looks like we can confirm that for a "simple" network, Triton is now indeed seeing the speedup of batches that one would expect. 

But.... it still baffles me... why would the Sentence-piece tokeniser from tensorflow text be slow when called from Triton, but fast when called from Tensorflow Serving? Isn't it the same Sentencepiece-tokeniser as is being used in Tensorflow serving? Is Triton doing something special with it?

Anyway, I will slowly build up the dummy network now with the pieces we have in the "large" network and see where it starts misbehaving.

**Comment by NiklasA11** [comment]
Confirmed that the problem was with a Sentencepiece tokenization layer, that for some reason worked well in Tensorflow Serving but for some still unknown reason did not work well together with Triton.

Using tfm.nlp.layers.SentencepieceTokenizer instead worked like a charm, so probably some minor implementation detail somewhere, but I am not going to dig into that. I'm just happy I have a working solution. 

Thanks for you help!

Best regards,
Niklas
